### PR TITLE
refactor!: TraceAPI

### PR DESCRIPTION
**Pull request by antazoey**
### What I did

fixes: #1587 
Fixes: APE-1389

Wanna try other providers? These ones are ready:

https://github.com/ApeWorX/ape-foundry/pull/85
https://github.com/ApeWorX/ape-alchemy/pull/63

### Checklist

- [ ] All changes are completed
- [ ] New test cases have been added
- [ ] Documentation has been updated

### CODE REVIEW

File: src/ape_ethereum/trace.py [hunk 1]
```diff
@@ -0,0 +1,538 @@
...
+        ls_spacing = spacing * (depth + 1)
+        new_val += f"{ls_spacing}{value[idx]}"
+        if idx < num_values - 1:
+            new_val += ","
+
+        new_val += "\n"
+
+    new_val += spacing * depth
+    new_val += "]"
+    return new_val
```
Comment (fubuloubu) [inline src/ape_ethereum/trace.py:538]:
```suggestion
    return (
        "[\n"
        + f",\n{spacing * (depth + 1)}".join(str(v) for v in value)
        + f"\n{spacing * depth}]"
    )
```

Comment (antazoey) [reply src/ape_ethereum/trace.py:556]: Thanks! I had to make 1 slight adjustment for accuracy:

```python
  spacing = _INDENT * " "
  ls_spacing = spacing * (depth + 1)
  joined = ",\n".join([f"{ls_spacing}{v}" for v in value])
  new_val = f"[\n{joined}\n{spacing * depth}]"
  return new_val
```

File: docs/userguides/trace.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+
+If you want to see the trace of call when making the call, use the `show_trace=` flag:
+
+```python
+token.balanceOf(account, show_trace=True)
+```
+
+**WARN**: If your provider does not properly support call-tracing (e.g. doesn't implement `debug_traceCall`), traces are limited to the top-level call.
+
+Ape traces calls automatically when using `--gas` or `--coverage` in tests to build reports.
```
Comment (fubuloubu) [inline docs/userguides/trace.md:87]: Wonder ifbthe coverage can use this method to remove the need to do any snapshot/reverts to speed things up a lot

**Review by fubuloubu — APPROVED** [review]

File: docs/userguides/trace.md (hunk 2 above)
Comment (antazoey) [reply docs/userguides/trace.md:87]: I am confused by this comment. Snapshoting and reverting is done to support test-isolation.
Test isolation is not required for --gas or --coverage to work.

Comment (fubuloubu) [reply docs/userguides/trace.md:87]: Don't we measure coverage for view functions?

**Review by fubuloubu — DISMISSED** [review]

File: src/ape/api/providers.py [hunk 3]
```diff
@@ -251,6 +243,30 @@ def network_choice(self) -> str:
...
+    @raises_not_implemented
+    def stream_request(  # type: ignore[empty-body]
+        self, method: str, params: List, iter_path: str = "result.item"
+    ) -> Iterator[Any]:
+        """
+        Stream a request, great for large requests like events or traces.
+
+        Args:
+            method (str): The RPC method to call.
+            params (List): Parameters for the method.s
```
Comment (fubuloubu) [inline src/ape/api/providers.py:263]: Can this relax to `Iterable`?

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**